cloud/aws/aws_s3.cc

```diff
@@ -818,14 +819,80 @@ Status S3StorageProvider::CopyCloudObject(const std::string& bucket_name_src,
   return st;
 }
 
+namespace {
+class WritableFileStreamBuf : public std::streambuf {
+ public:
+  WritableFileStreamBuf(std::unique_ptr<WritableFileWriter> fileWriter)
+      : fileWriter_(std::move(fileWriter)) {}
+
+  ~WritableFileStreamBuf() {
+    fileWriter_->Flush();
+    fileWriter_->Close();
+  }
+
+ protected:
+  std::streamsize xsputn(const char* s, std::streamsize n) override {
+    auto st = fileWriter_->Append(rocksdb::Slice(s, n));
+    if (!st.ok()) {
+      return EOF;
+    }
+    return n;
+  }
+
+  int_type overflow(int_type ch) override {
```

Comment:
also, it will be great to have some comments on each of these individual methods and what they are doing.